text/0031-corejam.md

```diff
@@ -0,0 +1,643 @@
+
+The function signature to the `accumulate` entry-point in the Work Class's code blob is:
+
+```rust
+fn accumulate(results: Vec<(Authorization, Vec<(ItemHash, WorkResult)>)>);
+type ItemHash = [u8; 32];
+```
+
+The logic in `accumulate` may need to know how the various Work Items arrived into a processed Work Package. Since a Work Package could have multiple Work Items of the same Work Class, it makes sense to have a separate inner `Vec` for Work Items sharing the Authorization (by virtue of being in the same Work Package).
+
+Work Items are identified by their Blake2-256 hash, known at the *Item Hash* (`ItemHash`). We provide both the Authorization of the Package and the constituent Work Item Hashes and their Results in order to allow the `refine` logic to take appropriate action in the case that an invalid Work Item was submitted (i.e. one which caused its Refine operation to panic or time-out).
+
+_(Note for later: We may wish to provide a more light-client friendly Work Item identifier than a simple hash; perhaps a Merkle root of equal-size segments.)_
+
+There is an amount of weight which it is allowed to use before being forcibly terminated and any non-committed state changes lost. The lowest amount of weight provided to `accumulate` is defined as the number of `WorkResult` values passed in `results` to `accumulate` multiplied by the `accumulate` field of the Work Class's weight requirements.
+
+However, the actual amount of weight may be substantially more. Each Work Package is allotted a specific amount of weight for all on-chain activity (`weight_per_package` above) and has a weight liability defined by the weight requirements of all Work Items it contains (`total_weight_requirement` above). Any weight remaining after the liability (i.e. `weight_per_package - total_weight_requirement`) may be apportioned to the Work Classes of Items within the Report on a pro-rata basis according to the amount of weight they utilized during `refine`. Any weight unutilized by Classes within one Package may be carried over to the next Package and utilized there.
+
+```rust
+fn get_work_storage(key: &[u8]) -> Option<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]) -> Option<u32>;
+fn checkpoint() -> Weight;
+fn weight_remaining() -> Weight;
+fn set_work_storage(key: &[u8], value: &[u8]) -> Result<(), ()>;
+fn remove_work_storage(key: &[u8]);
+fn set_validators(validator_keys: &[ValidatorKey]) -> Result<(), ()>;
+fn set_code(code: &[u8]) -> Result<(), ()>;
+fn assign_core(
+    core: CoreIndex,
+    begin: BlockNumber,
+    assignment: Vec<(CoreAssignment, PartsOf57600)>,
+    end_hint: Option<BlockNumber>,
+) -> Result<(), ()>;
+fn transfer(
+    destination: WorkClass,
+    amount: u128,
+    memo: &[u8],
+    weight: Weight,
+) -> Result<Vec<u8>, ()>;
```

Comment:
Some of these host functions are clashing with the idea in https://github.com/polkadot-fellows/RFCs/pull/4.

All the functions that return a `Result` should ideally return a number. While this seems a no-brainer, if someone just copy-pastes these definitions in Substrate, the code generated by Substrate will instead use the allocator just to allocate one byte and write 0 or 1 in it.

I would suggest turning `get_work_storage` into `read_work_storage`, where a pointer to buffer and a maximum length are passed as parameter.

As for `transfer`, I didn't figure out what is being returned, so don't have an opinion.